lib/charms/loki_k8s/v0/loki_push_api.py

```diff
@@ -1250,29 +1289,40 @@ def _process_logging_relation_changed(self, relation: Relation):
         if self._charm.unit.is_leader():
             relation.data[self._charm.app].update(self._promtail_binary_url)
             logger.debug("Saved promtail binary url: %s", self._promtail_binary_url)
-            relation.data[self._charm.app]["endpoints"] = json.dumps(self._endpoints())
-            logger.debug("Saved endpoints in relation data")
 
         if relation.data.get(relation.app).get("alert_rules"):
             logger.debug("Saved alerts rules to disk")
             self._remove_alert_rules_files(self.container)
             self._generate_alert_rules_files(self.container)
             self._check_alert_rules()
 
-    def _endpoints(self) -> List[dict]:
-        """Return a list of Loki Push Api endpoints."""
-        return [{"url": self._url(unit_number=i)} for i in range(self._charm.app.planned_units())]
-
-    def _url(self, unit_number) -> str:
-        """Get the url for a given unit."""
-        return "http://{}-{}.{}-endpoints.{}.svc.cluster.local:{}/loki/api/v1/push".format(
-            self._charm.app.name,
-            unit_number,
-            self._charm.app.name,
-            self._charm.model.name,
-            self.port,
+    def _set_endpoint_on_relation(self, relation: Relation) -> None:
+        """Set the the loki_push_api endpoint on relations.
+
+        Args:
+            relation: Only this relation will be updated.
+                Otherwise, all instances of the `prometheus_remote_write`
+                relation managed by this `PrometheusRemoteWriteProvider` will be
+                updated.
+        """
+        address = self._endpoint_address or self._get_relation_bind_address()
+        path = self._endpoint_path or ""
+
+        if path and not path.startswith("/"):
+            path = "/{}".format(path)
+
+        endpoint_url = "{}://{}:{}{}".format(
+            self._endpoint_schema, address, str(self._endpoint_port), path
+        )
+        relation.data[self._charm.unit]["endpoint"] = json.dumps({"url": endpoint_url})
+        logger.debug(
+            "Saved endpoint %s in relation data. ", relation.data[self._charm.unit]["endpoint"]
         )
 
+    def _get_relation_bind_address(self):
+        network_binding = self._charm.model.get_binding(self._relation_name)
```

Comment:
Note that this may return None. Not sure if relevant for the circumstances here.
https://bugs.launchpad.net/juju/+bug/1929364